### ISSUE TITLE
OS* classes design

### ISSUE THREAD

**Issue by npcottaris** [isetbio/isetbio#97]
Hello,

Questions about the design of the OS* classes:

1. When the noiseFlag is set to 1, osCompute adds noise to the response ('coneCurrentSignal') but stores the result in a different variable ('coneCurrentSignalPlusNoise'). 

2. Moreover, one can only query (via osGet) only the value of  'coneCurrentSignal', as the 'coneCurrentSignalPlusNoise' is not in the list of allowable properties to get.

So currently, there is no way to obtain the noisy response. But even if one were to add the 'coneCurrentSignalPlusNoise' to the list of allowable getable properties, I do not think this is the design we want. I think that when the user asks for a noisy response, he should not have to query for a different property than when he is asking for a noise-free response.

These issues  can both be fixed by saving the noisy response to 'coneCurrentSignal' and by eliminating the 'coneCurrentSignalPlusNoise' property. Is there a reason why we are not presently doing so?

Nicolas

**Comment by fmrieke** [comment]
> These issues can both be fixed by saving the noisy response to 'coneCurrentSignal' and by eliminating the 'coneCurrentSignalPlusNoise' property. Is there a reason why we are not presently doing so?
> 
> 
This seems like the right thing to do.  I can imagine cases where you would want both, but you could always run with and without the noise flag.  If that seems too time consuming (so far other factors have been slower than the cone model itself) we could make both available.  

Fred

**Comment by jamesgolden1** [comment]
Thank you for the comments Nicolas and Fred - I will change osGet/Set to
make the noisy signal available. I decided to keep both because I have had
use for both.

Another issue is that we will likely want some number of noisy versions of
the signal. In the HLColorGaborOpponent tutorial, I compute 100 copies of
the noisy signal, and since the signal is a large (x,y,t) matrix of cone
current values, generating 100 copies takes several minutes.

Would it be worth incorporating into the OS object a property for number of
noisy repeats? Presently I store the actual noisy signals in their own
variable, but they could also be stored in the object, although of course
for large N this would greatly increase the memory taken up by the object
variable itself.

**Comment by fmrieke** [comment]
This seems like a general issue that applies to the os object, but will apply to others as well.  It seems best to store the simulated responses directly in the object for bookkeeping reasons.  Could we include (for this and other objects) a flush command or something similar that would clear out memory as needed?

Fred

**Comment by jamesgolden1** [comment]
I removed the property "coneCurrentSignalPlusNoise", so now there is only "coneCurrentSignal", and the noise flag determines whether or not it has noise.

I have not altered object to store N noisy copies yet. I will work on that later this week along with the HLGaborColorOpponent tutorial.

**Comment by DavidBrainard** [comment]
I thought we concluded that we do not want to store N noisy copies.  Rather, we want the ability for the object to generate N copies of noise and return (but not store internally) N noisy copies, based on
the stored noise free response.

Best,

David

On Jan 5, 2016, at 2:42 PM, jamesgolden1 <notifications@github.com<mailto:notifications@github.com>> wrote:


I removed the property "coneCurrentSignalPlusNoise", so now there is only "coneCurrentSignal", and the noise flag determines whether or not it has noise.

I have not altered object to store N noisy copies yet. I will work on that later this week along with the HLGaborColorOpponent tutorial.

**Comment by fmrieke** [comment]
It seems like this depends entirely on performance issues.  I agree that generating the noise as needed is ideal, but if that proves rate-limiting for some calculations it would be better to store the N noise copies.  Maybe we should run some benchmarks on performance with and without noise generation and could make a decision based on that?

Fred

**Comment by jamesgolden1** [comment]
I timed the execution of t_colorGaborDetection.m (in the HLColorGaborOpponent repository of isetbio). This is a scaled-down simulation of Fred's paper with Greg Horwitz (et al.). Given the size of the stimulus and the eccentricity at which it is presented, the sensor array turns out to be rather small at 17x20 cones. Following the experiment, the sampling interval is 1 ms, and there are 666 time steps for that result in an os with the coneCurrentSignal matrix at 17x20x666. 

By far the slowest part of this is the generation of the 666 frames of the dynamic stimulus. This takes about 6 minutes (and could be much longer for a larger sensor array). The duration of the osCompute step for the noise-free signal is negligible (0.04 sec). To generate 500 noisy versions of the signal takes 60 seconds. However, following the paper, the simulation uses an ideal observer/matched filter to take each 17x20x666 noisy signal and project it down to a vector with three values representing the pooled activity of each cone type. Only these three values are stored for each noisy signal, so there is clearly no memory issue for this simulation.

In order to see the memory requirements if we did want to store 500 versions of the noisy signal, I created a variable with zeros(500,17,20,666), which is about 900 MB as a double. If the noisy os array were larger, like zeros(500,40,40,666), that is 4.3 GB. Of course this will be different for every simulated experiment, and we can change this with the number of noisy repeats, but this makes me wary of storing the noisy versions of the signal internally to the object where a user might not have a sense of how much memory the noisy signals take up.

Anyway, the take-away is that dynamic scenes take a long time to generate. The noisy os signals are generated quickly in comparison, but storing each is in my opinion not feasible because of possible memory issues.

**Comment by fmrieke** [comment]
I agree - that seems pretty clear cut that we should regenerate the responses whenever we need them (unless we identify some other application where generating the responses is limiting).  

Fred
